File: Data Structures & Algorithms/matrixBFS/submission-0.py
class Solution:
    def shortestPath(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])

        dis = [[-1]*COLS for _ in range(ROWS)]
        dis[0][0] = 0

        q = []
        q.append((0,0))

        way = [(1,0), (-1,0), (0,1), (0,-1)]

        while len(q):
            r,c = q.pop()
            print(f"r {r} - c {c}")

            for dx,dy in way:
                if min(r+dx, c+dy)>=0 and r+dx<ROWS and c+dy<COLS and grid[r+dx][c+dy]!=1 and (dis[r+dx][c+dy]==-1 or dis[r+dx][c+dy]>dis[r][c]+1):
                    q.append((r+dx,c+dy))
                    dis[r+dx][c+dy] = dis[r][c]+1
            

        return dis[ROWS-1][COLS-1]
            

        

File: Data Structures & Algorithms/islands-and-treasure/submission-0.cpp
class Solution {
public:
    void islandsAndTreasure(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();

        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(grid[i][j]==0){
                    bfs(i, j, n, m, grid);
                }
            }
        }

        // for(int i=0; i<n; i++){
        //     for(int j=0; j<m; j++){
        //         cout<<grid[i][j]<<" ";
        //     } cout<<endl;
        // }
    }

    void bfs(int i, int j, int n, int m, vector<vector<int>>& grid){
        queue<vector<int>>q;
        q.push({i,j,0});

        while(!q.empty()){
            vector<int> u = q.front();
            q.pop();
            int x = u[0];
            int y = u[1];
            int dis = u[2];

            if(x-1>=0 && grid[x-1][y]!=0 && grid[x-1][y]!=-1 && dis+1<grid[x-1][y]){
                grid[x-1][y] = dis + 1;
                q.push({x-1,y,dis+1});
            }
            if(x+1<n && grid[x+1][y]!=0 && grid[x+1][y]!=-1 && dis+1<grid[x+1][y]){
                grid[x+1][y] = dis + 1;
                q.push({x+1,y,dis+1});
            }
            if(y-1>=0 && grid[x][y-1]!=0 && grid[x][y-1]!=-1 && dis+1<grid[x][y-1]){
                grid[x][y-1] = dis + 1;
                q.push({x,y-1,dis+1});
            }
            if(y+1<m && grid[x][y+1]!=0 && grid[x][y+1]!=-1 && dis<grid[x][y+1]){
                grid[x][y+1] = dis + 1;
                q.push({x,y+1,dis+1});
            }
        }
    }
};
